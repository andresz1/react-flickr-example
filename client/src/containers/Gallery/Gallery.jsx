// @flow

import React, { Component } from 'react';
import InfiniteScroll from 'react-infinite-scroller';
import { connect } from 'react-redux';
import { RotateSpinLoader } from 'react-css-loaders';

import PictureGrid from '../../components/PictureGrid';
import PictureModal from '../../components/PictureModal';
import { fetchPictures, selectPicture } from '../../actions/picturesActions.js';
import type { IPicturesState } from '../../reducers/picturesReducer.js';

type IProps = {
  dispatch: Function,
  pictures: IPicturesState,
};

class Gallery extends Component<IProps> {
  handleMore() {
    const { dispatch, pictures } = this.props;

    if (!pictures.isFetching) {
      dispatch(fetchPictures(pictures.page + 1));
    }
  }

  handleSelect(picture: IPicture)Â {
    const { dispatch } = this.props;

    dispatch(selectPicture(picture));
  }

  handleNext() {
    const { dispatch, pictures } = this.props;

    let index: number = pictures.photo.indexOf(pictures.selected) + 1;
    index = index % pictures.photo.length;

    dispatch(selectPicture(pictures.photo[index]));
  }

  handlePrevious() {
    const { dispatch, pictures } = this.props;

    let index: number = pictures.photo.indexOf(pictures.selected) - 1;
    index = (index < 0) ? pictures.photo.length - 1 : index;

    dispatch(selectPicture(pictures.photo[index]));
  }

  handleClose() {
    const { dispatch } = this.props;

    dispatch(selectPicture());
  }

  renderModal() {
    const { pictures } = this.props;

    if (pictures.selected) {
      return (
        <PictureModal
          picture={pictures.selected}
          onNext={this.handleNext.bind(this)}
          onPrevious={this.handlePrevious.bind(this)}
          onClose={this.handleClose.bind(this)}
        />
      );
    }
  }

  render() {
    const { pictures } = this.props;

    const hasMore : boolean = pictures.pages === undefined || pictures.page < pictures.pages;

    return (
      <InfiniteScroll
        loadMore={this.handleMore.bind(this)}
        hasMore={hasMore}
        loader={
          <RotateSpinLoader
            color="#FF0084"
            size="3"
          />
        }
      >
        <PictureGrid
          pictures={pictures.photo}
          onSelect={this.handleSelect.bind(this)}
        />

        {this.renderModal()}
      </InfiniteScroll>
    );
  }
}

const mapStateToProps = (state) => ({
  pictures: state.pictures,
});

export default connect(mapStateToProps)(Gallery);
