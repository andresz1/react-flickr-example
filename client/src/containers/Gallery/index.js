import Gallery from './Gallery.jsx';

export default Gallery;
