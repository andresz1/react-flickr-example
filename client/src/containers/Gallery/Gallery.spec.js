import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk'
import { shallow } from 'enzyme';

import Gallery from './Gallery';

const picture = {
  id: '44592167171',
  owner: '166588675@N05',
  secret: '144e4f8561',
  server: '1848',
  farm: 2,
  title: 'Olden 3',
  ispublic: 1,
  isfriend: 0,
  isfamily: 0,
  ownername: 'OldenCam3',
  tags: '',
  src: 'https://farm2.staticflickr.com/1848/44592167171_144e4f8561.jpg',
  href: 'https://www.flickr.com/photos/166588675@N05/44592167171/'
};

const setup = (setupProps = {}) => {
  const middlewares = [ thunk ];
  const store = configureStore(middlewares)({
    pictures: {
      photo: [picture],
      page: 0,
      pages: 5
    }
  });
  const wrapper = shallow(<Gallery store={store} />);

  return {
    store,
    wrapper
  };
};

describe('<Gallery />', () => {
  it('renders without crashing', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('handles pictures fetching', () => {
    const { store, wrapper } = setup();

    wrapper.dive().instance().handleMore();
    expect(store.getActions()).toEqual([{
      type: 'FETCH_PICTURES_REQUEST',
      payload: {},
    }]);
  });

  it('handles picture selection', () => {
    const { store, wrapper } = setup();

    wrapper.dive().instance().handleSelect(picture);
    expect(store.getActions()).toEqual([{
      type: 'SELECT_PICTURE',
      payload: picture,
    }]);
  });

  it('handles previous picture', () => {
    const { store, wrapper } = setup();

    wrapper.dive().instance().handlePrevious();
    expect(store.getActions()).toEqual([{
      type: 'SELECT_PICTURE',
      payload: picture,
    }]);
  });

  it('handles next picture', () => {
    const { store, wrapper } = setup();

    wrapper.dive().instance().handleNext();
    expect(store.getActions()).toEqual([{
      type: 'SELECT_PICTURE',
      payload: picture,
    }]);
  });
});
