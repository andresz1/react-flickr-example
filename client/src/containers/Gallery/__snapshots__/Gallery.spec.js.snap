// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gallery /> renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Gallery)
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "dispatch": [Function],
      "pictures": Object {
        "page": 0,
        "pages": 5,
        "photo": Array [
          Object {
            "farm": 2,
            "href": "https://www.flickr.com/photos/166588675@N05/44592167171/",
            "id": "44592167171",
            "isfamily": 0,
            "isfriend": 0,
            "ispublic": 1,
            "owner": "166588675@N05",
            "ownername": "OldenCam3",
            "secret": "144e4f8561",
            "server": "1848",
            "src": "https://farm2.staticflickr.com/1848/44592167171_144e4f8561.jpg",
            "tags": "",
            "title": "Olden 3",
          },
        ],
      },
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "dispatch": [Function],
        "pictures": Object {
          "page": 0,
          "pages": 5,
          "photo": Array [
            Object {
              "farm": 2,
              "href": "https://www.flickr.com/photos/166588675@N05/44592167171/",
              "id": "44592167171",
              "isfamily": 0,
              "isfriend": 0,
              "ispublic": 1,
              "owner": "166588675@N05",
              "ownername": "OldenCam3",
              "secret": "144e4f8561",
              "server": "1848",
              "src": "https://farm2.staticflickr.com/1848/44592167171_144e4f8561.jpg",
              "tags": "",
              "title": "Olden 3",
            },
          ],
        },
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
